### DISCUSSION TITLE
`env` function is not returning the value of the environment variable

### DISCUSSION THREAD

**Discussion by houssemz** [symfony/symfony#49039] [Q&A]
I am working on a project that uses many env vars and I used php files (rather than yaml files or xml) to deal with configurations.

I am trying to this code :

```php 
    // config/my_conf_name.php
    
    return static function (ContainerConfigurator $containerConfigurator): void {
       $containerConfigurator->extension('my_extension', [
           ...
       ]);
    
        // CLIENT_TYPE is an env variable
        if ('http' === (string) env('CLIENT_TYPE')->string()) {
            $services->alias(AbstractClient::class, HttpClient::class);
        } else {
            $services->alias(AbstractClient::class, AwsClient::class);
        }
        
        ...
    }
```

Unfortunately, `env('CLIENT_TYPE')->string())` isn’t parsed

```php
dump(
    (string) env('CLIENT_TYPE')->string())
)

// output
^ "%env(string:CLIENT_TYPE)%"
```

Obviously, when I do this, it works for me :

```php
    if (isset($_ENV['CLIENT_TYPE']) && 'http' === $_ENV['CLIENT_TYPE']) {
       ....
    }
```

It looks like the issue is that the env function is not returning the value of the environment variable, but rather the string representation of the variable.

**Comment by houssemz** [comment]
@nicolas-grekas Am I wrong?

**Reply by nicolas-grekas** [reply]
This is a support request yes as it looks like you didn't fully grasp the concept of env vars. I don't know where it's written explicitly enough (if it's not, it might be great to improve the doc, PR welcome), but the very nature of env vars is for them to be read *at runtime*. But a config file is a compile-time piece of logic, so reading env vars there breaks their dynamic nature.

**Reply by houssemz** [reply]
Hmm okay 
In my case, I need to read the env file at the runtime. Is there a better option rather the doing this is a config file?
Thanks!

**Reply by nicolas-grekas** [reply]
I guess you need a factory service here, that will take the env as argument and will return the service you want based on its value at runtime.
